syncer/syncer.go

```diff
@@ -632,7 +628,7 @@ func (s *Syncer) syncLoop() error {
 // Run spawns goroutines for accepting inbound connections, forming outbound
 // connections, and syncing the blockchain from active peers. It blocks until an
 // error occurs, upon which all connections are closed and goroutines are
-// terminated. To gracefully shutdown a Syncer, close its net.Listener.
+// terminated.
 func (s *Syncer) Run() error {
```

Comment:
I think you can get rid of the `shutdownCtx` and `shutdownCtxCancel` fields by moving it here. Has, essentially, the same effect now that the waitgroup is there and is more idiomatic.

```go
func (s *Syncer) Run(ctx context.Context) error {
	errChan := make(chan error)
	go func() { errChan <- s.acceptLoop(ctx) }()
	go func() { errChan <- s.peerLoop(ctx) }()
	go func() { errChan <- s.syncLoop(ctx) }()
	err := <-errChan
        // ....
}
```